basic PBR example

devdad opened the pull request:
pbr implementation  includes rpbr.h and few shader files header only file, which self contain everything needed for pbr rendering. Few textures and one model of the car which is under free licence which is included inside basic_pbr.c example file  currently supported shader versions are 120 and 330 , version 100 has small issue which I have to resolve

victorfisac commented:
Thank you for this effort

devdad commented:
No problem Ray! It was my pleasure.  Please ping me if you need anything.

devdad commented:
> Thank you for this effort

Thank you Victor! It was much easier because of you!

devdad commented on examples/shaders/rpbr.h:
```diff
@@ -0,0 +1,458 @@
+
+void PBRSetVec2(PBRMaterial *pbrMat,PBRVec2Type type,Vector2 value){
+    pbrMat->texTiling[0] =value.x;
+    pbrMat->texTiling[1] =value.y;
```
Ahhh good one, it was value was Vector4 at first this is just a mistake yeah! Thanks

devdad replied:
Yeah, good one!  The value was  Vector4 at first so it is just a mistake.  Thanks !  I will change that!

devdad replied:
Ok done, thanks again!

TokyoSU replied:
No problem :)

raysan5 commented:
@devdad Is the `plane.glb` really needed? Could it be created with raylib directly using `GenMeshPlane()`?

raysan5 commented:
How big are the textures included in pixels? I see they add about 15MB to the repo, that's quite big... maybe texture size could be reduced to 1/4, at least some of them?

devdad commented:
Hey Ray we can reduce 1/2 and save as jpg so it would be good quality and smaller?

devdad commented:
plane.glb really needed?  not really , but if we use the GenMeshPlane() we just need to create Tangents

raysan5 commented:
> Hey Ray can we reduce 1/2 and save it as jpg so it would be good quality and smaller?

If width/2 and height/2 it would be reduced to 1/4, JPG would lose image quality so it's ok to keep them as PNG. Considering it would be used for other examples, maybe only some maps of the car/road can be reduced? It would be nice to make some quick test to see if it is really noticeable the change... Also note that this is an example for the users to see how to implement it, finally visual quality is not that important.

 > "plane.glb really needed?" Not really, but if we use the GenMeshPlane() we just need to create tangents, it is even better because we will show that we need to generate tangents, otherwise people will be confused why it is not working.

That's fine, you can use `GenMeshTangents()`.

raysan5 commented on examples/shaders/basic_pbr.c:
```diff
@@ -0,0 +1,170 @@
+/*******************************************************************************************
+*
+*   raylib [core] example - Model Defuse Normal Shader (adapted for HTML5 platform)
```
Example file name should be `shaders_basic_pbr.c` for consitency with the other shaders examples.

devdad replied:
ok

devdad commented:
yeah agree I will test. I think it would be still good .

raysan5 commented:
@devdad I'm checking `rpbr.h` and I really think it can directly use raylib `Model` and `Material` structures but the refactor could be considerable so I will merge it (after the textures size review/naming) and I'll try to work on it myself (also to validate if it is really possible or some parameter needs to be added to raylib structures).

Also noted that at the moment there is no `PBREnvironment` in use, right?

devdad commented:
no, for now there is no PBREnvironment but that will come soon IBL, reflection etc. we can use raylib Model and Material but we need to change them a bit, my intention was not to change anything because you already have a lot to do, and PBR can have just about anything in the header, especially since it will be expanded soon
with all the goodies, so maybe it makes sense =) But it's a good idea to at least update the texture enumerations with MRA_TEXTURE and location, because it's really a common implementation to have rgb metallic , roughness , and ao together in one texture.

devdad commented:
hey Ray, I di change everything , I changed texture size to 1024x1024  from 2048x2048 , changed the file name to shaders_basic_pbr.c   added function  PBRLoadModelFromMesh(Mesh mesh);  But I couldn't use  GenMeshPlane(2.0,2.0,3,3);   because plane crash the function GenMeshTangents();  I didn't have time to investigate why, but I did use before GenMeshTangents()  on Loaded meshes and it was working.

raysan5 commented:
@devdad thank you very much for the addition and the reviews, I'm merging it for further reviews.
